maze_transformer/generation/generators.py

```diff
@@ -106,7 +106,100 @@ def gen_dfs_with_solution(cls, grid_shape: Coord):
+        """Generate a lattice maze using Wilson's algorithm. Wilson's algorithm generates an unbiased sample maze
+        from the uniform distribution over all mazes, using loop-erased random walks.
+        https://en.wikipedia.org/wiki/Maze_generation_algorithm#Wilson's_algorithm"""
+
+        def neighbor(current: Coord, direction: int) -> Coord:
+            x, y = current
+
+            if direction == 0:
+                y -= 1  # Left
+            elif direction == 1:
+                y += 1  # Right
+            elif direction == 2:
+                x -= 1  # Up
+            elif direction == 3:
+                x += 1  # Down
+            else:
+                return None
+
+            return np.array([x, y]) if 0 <= x < height and 0 <= y < width else None
+
+        height, width = grid_shape
+
+        # A connection list only contains two elements: one boolean matrix indicating all the
+        # downwards connections in the maze, and one boolean matrix indicating the rightwards connections.
+        connection_list: np.ndarray = np.zeros((2, height, width), dtype=bool)
+
+        connected = np.zeros(grid_shape, dtype=bool)
+        direction_matrix = np.zeros(grid_shape, dtype=int)
+
+        # Mark a random cell as connected
+        connected[random.randint(0, height - 1)][random.randint(0, width - 1)] = True
+
+        cells_left: int = height * width - 1
+        while cells_left > 0:
+            current: Coord = np.array(
+                [random.randint(0, height - 1), random.randint(0, width - 1)]
+            )
+            start: Coord = current
+
+            # Random walk through the maze until a connected cell is found
+            while not connected[current[0]][current[1]]:
+                # Find a valid neighboring cell by checking in a random direction then rotating clockwise
+                direction: int = random.randint(0, 4)
+                next: Coord = neighbor(current, direction)
+
+                while next is None:
+                    direction += 1
+                    if direction > 3:
+                        direction = 0
+                    next = neighbor(current, direction)
+
+                # Keep track of the random path
+                direction_matrix[current[0]][current[1]] = direction
+                # Move to the neighboring cell
+                current = next
+
+            direction_matrix[current[0]][current[1]] = 4
+
+            # Return to the start and retrace our path, connecting cells as we go
+            current = start
+            while not connected[current[0]][current[1]]:
+                direction = direction_matrix[current[0]][current[1]]
+                connected[current[0]][current[1]] = True
+                cells_left -= 1
+
+                next = neighbor(current, direction)
+                # Connect the current and next cell
+                # todo(luciaq) update LatticeMaze to take an adjacency list instead of a connection list for a more
```

Comment:
I'm not quite sure exactly what this TODO entails, but doesn't `LatticeMaze` have a `from_adj_list()` classmethod?

Reply:
Yep it does! I was thinking of adding a LatticeMaze initialisation method that takes an adjacency list instead of a connection list because it's the standard data structure used for most maze generation implementations I've seen, but after coding this generation algorithm up it doesn't seem worth it anymore, the data structures are quite similar

Reply:
are you saying the constructor should be changed to also accept the adjacency list instead of having that only be in a separate method?